google/cloud/forseti/common/opencensus/tracing.py

```diff
@@ -0,0 +1,349 @@
+        # Adds `self.tracer` in class to give access to the tracer from within.
+        if OPENCENSUS_ENABLED:
+            for name, func in to_trace:
+                LOGGER.debug(f'Tracing - Adding decorator to {name}')
+                if name == '__init__':
+                    # __init__ decorator to add tracer as instance attribute
+                    decorator = trace_init(attr=attr)
+                else:
+                    decorator = trace()
+                setattr(cls, name, decorator(func))
+        return cls
+    return wrapper
+
+
+# pylint: disable=redundant-returns-doc
+def trace_init(attr=None):
+    """Method decorator for a class's __init__ method.
+    Set `self.tracer` (either from instance kwargs, attribute,
+    or execution context).
+
+    Args:
+        attr (str, optional): If the tracer was passed explicitly to the class
+            in an attribute, get it from there.
+
+    Returns:
+        func: Decorated function.
+    """
+    def outer_wrapper(init):
+        """Outer function.
+
+        Args:
+            init (method): Initialization method.
+
+        Returns:
+            func: A function that is decorated.
+        """
+        @functools.wraps(init)
+        def inner_wrapper(self, *args, **kwargs):
+            """Inner function.
+
+            Args:
+                *args: Arguments of __init__ method.
+                **kwargs (dict): keyword arguments of __init__ method.
+
+            Returns:
+                func: Method decorator for a class's __init__ method.
+            """
+            cls_name = self.__class__.__name__
+            LOGGER.debug(f'Decorating {cls_name}')
+            init(self, *args, **kwargs)
+
+            if OPENCENSUS_ENABLED:
+                if 'tracer' in kwargs:
+                    # If `tracer` is passed explicitly to our class at __init__,
+                    # we use that tracer.
+                    LOGGER.debug(f'Tracing - {cls_name}.__init__ - set tracer '
+                                 f'from kwargs')
+                    self.tracer = kwargs['tracer']
+                elif attr is not None:
+                    # If `attr` is passed to this decorator, then get the tracer
+                    # from the instance attribute.
+                    LOGGER.debug(f'Tracing - {cls_name}.__init__ - set tracer '
+                                 f'from class attribute {cls_name}')
+                    self.tracer = rgetattr(self, cls_name)
+                else:
+                    # Otherwise, get tracer from current execution context.
+                    LOGGER.debug(f'Tracing - {cls_name}.__init__ - set tracer '
+                                 f'from execution context')
+                    self.tracer = execution_context.get_opencensus_tracer()
+                LOGGER.debug(f'Tracing - {cls_name}.__init__ - '
+                             f'context: {self.tracer.span_context}')
+        return inner_wrapper
+    return outer_wrapper
+
+
+def trace():
```

Comment:
@c24t could you please review this decorator's code ? Trying to find something in there that would explain why `trace_id` for successive run is the same and does not seem to be reset. Thanks :)